crates/service/src/algorithms/flat.rs

```diff
@@ -56,64 +45,80 @@ impl<S: G> Flat<S> {
 
     pub fn vbase<'a>(
         &'a self,
-        vector: &'a [S::Scalar],
+        vector: &'a [S::Element],
         _opts: &'a SearchOptions,
         filter: impl Filter + 'a,
     ) -> (Vec<Element>, Box<(dyn Iterator<Item = Element> + 'a)>) {
         vbase(&self.mmap, vector, filter)
     }
+
+    pub fn dims(&self) -> u16 {
+        self.mmap.raw.dims()
+    }
+
+    pub fn len(&self) -> u32 {
+        self.mmap.raw.len()
+    }
+
+    pub fn content(&self, i: u32) -> &[S::Element] {
```

Comment:
Too. We can add a associcated type `vector<'a>` to `G`.

Reply:
There is a function to expand a sparse vector.

https://github.com/silver-ymz/pgvecto.rs/blob/feat/sparse/crates/service/src/prelude/storage/mod.rs#L32

Reply:
I prefer such a structure:
```rust
struct Vecf32<'a> {
    dimensions: &'a [F32],
}
struct SVecf32<'a> {
    dims: u16,
    elements: &'a [(u16, F32)],
}
```
It will make programming easier.

Reply:
added in new commit. But I think it seems to introduce more intermediate types, making programming more complex.